zerver/lib/actions.py

```diff
@@ -3926,24 +3927,29 @@ def __init__(self, msg: Text, errors: List[Tuple[Text, str]], sent_invitations:
         self.errors = errors  # type: List[Tuple[Text, str]]
         self.sent_invitations = sent_invitations  # type: bool
 
-def do_invite_users(user_profile: UserProfile,
-                    invitee_emails: SizedTextIterable,
-                    streams: Iterable[Stream],
-                    invite_as_admin: Optional[bool]=False,
-                    body: Optional[str]=None) -> None:
+def check_invite_limit(user: UserProfile, num_invitees: int) -> None:
     if settings.OPEN_REALM_CREATION:
         # Discourage using invitation emails as a vector for carrying spam
-        sent_invites = Confirmation.objects.filter(
-            realm=user_profile.realm,
-            date_sent__gte=timezone_now() - datetime.timedelta(days=1),
-            type=Confirmation.INVITATION).count()
-        if len(invitee_emails) + sent_invites > user_profile.realm.max_invites:
+        sent_invites = RealmCount.objects.filter(
+            realm=user.realm,
+            property='invites_sent::day',
+            end_time__gte=timezone_now() - datetime.timedelta(days=1)
+        ).aggregate(Sum('value'))['value__sum']
```

Comment:
This will include yesterday's count, but not today's. Right?

As that demonstrates, this logic is kind of tricky. It'd be best as its own function probably in `analytics.lib.counts`. Maybe like `count_upper_bound(stat, *, start_time=None, end_time=None)`? Then that function can take care of the boundary issues to get all the recorded intervals that intersect with the requested interval.

Reply:
the current code gets yesterday's + today's, right?

Reply:
I'm looking at the `end_time` condition... oh. Yeah, in fact I was reading it wrong. :-)

Still, I think it stands that this is pretty tricky logic and it'd be good to extract.

Reply:
Looking at this: I agree that if this were a way we wanted to access CountStats in the future, it would make sense to have a function for it in counts.py, since it is pretty tricky. But the general philosophy behind the analytics tables so far has been to expose a very limited set of ways of accessing the data, that are cheap (just reads, indexed, etc), and have all the computation done in the calling code or in the cron job.

It might make sense to change this for a certain category of stats (e.g. rate limiting stats, or performance measuring stats), but it is not the design I would choose if we were going to add more of these rate-limiting type things. So I'm not sure I want to cement this pattern further, especially given that this may not even be the final form of `check_invite_limit`.

Reply:
Hmm, I see.

In that case, perhaps just a specific function with this meaning, just above in this file? Like `upper_bound_recent_invites(timedelta)`. Just putting that description in, that this is (a) supposed to be inexact but in particular an upper bound, and (b) is supposed to cover the given interval (and not e.g. one day more or less), would I think help the reader think through the boundary issues -- they have your statement of the lemma, and then can separately check the lemma and/or look at how its conclusion is used.

Or a comment here explaining it, but I think that isn't actually less work than pulling this expression (and the None->0 below) into a function and giving it a name.